File: src/main/java/cat/inspiracio/calculator/WorldRepresentation.java
/*	Copyright 2011 Alexander Bunkenburg alex@inspiracio.cat
 * 
 * This file is part of Complex Calculator.
 * 
 * Complex Calculator is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Complex Calculator is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Complex Calculator. If not, see <http://www.gnu.org/licenses/>.
 * */
package cat.inspiracio.calculator;

import cat.inspiracio.complex.Complex;
import cat.inspiracio.complex.package$;
import cat.inspiracio.numbers.ECList;
import cat.inspiracio.numbers.Piclet;

import javax.swing.*;
import java.awt.*;

// Referenced classes of package bunkenba.calculator:
//            DoubleBuffer, World, Drawing

abstract class WorldRepresentation extends JComponent {

    /** Initial size wide enough so that the interaction menu is visible. */
    private static final Dimension MIN_SIZE = new Dimension(550, 300);

    //State ---------------------------------------------------------
	
    protected World w;
    protected DoubleBuffer doubleBuffer;

	//Constructors --------------------------------------------------
	
    protected WorldRepresentation(World world){
        w = world;
        doubleBuffer = new DoubleBuffer(this);
        setBackground(Color.white);
    }
    
    //Methods ------------------------------------------------------

    abstract void drawComplex(Drawing drawing, Complex ec);

    abstract void drawECList(Drawing drawing, ECList eclist);

    void drawPiclet(Drawing drawing, Piclet piclet){
        drawECList(drawing, piclet.getSamples());
    }

    @Override public Dimension getPreferredSize(){return getMinimumSize();}
    @Override public Dimension getMinimumSize(){return MIN_SIZE;}

    abstract void moveTo(Drawing drawing, Complex ec);

    abstract void lineTo(Drawing drawing, Complex ec);

    abstract Complex Point2Complex(Point point);

    abstract void reset();

    abstract void shift(int i, int j);

    abstract void zoomIn();

    abstract void zoomOut();

    // helpers -----------------------------------

    protected boolean finite(Complex z){return package$.MODULE$.finite(z);}
    protected double Im(Complex c){ return cat.inspiracio.complex.package$.MODULE$.Im(c); }
    protected double Re(Complex c) { return cat.inspiracio.complex.package$.MODULE$.Re(c); }
    protected double abs(Complex z){ return package$.MODULE$.abs(z); }

    protected String toString(double d){ return package$.MODULE$.double2Complex(d).toString(); }

    protected static final double sqr(double d){return d * d;}

    protected final Complex infinity = Real(1).$div(0);

    protected Complex Cartesian(double re, double im){
        Complex i = package$.MODULE$.i();
        return i.$times(im).$plus(re);
    }

    protected Complex Real(double re){ return package$.MODULE$.double2Complex(re); }

}

File: src/main/scala/cat/inspiracio/calculator/WorldRepresentation.scala
/*	Copyright 2011 Alexander Bunkenburg alex@inspiracio.cat
 *
 * This file is part of Complex Calculator.
 *
 * Complex Calculator is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Complex Calculator is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Complex Calculator. If not, see <http://www.gnu.org/licenses/>.
 * */
package cat.inspiracio.calculator

import java.awt.{Color, Dimension, Graphics, Point}

import cat.inspiracio.complex._
import cat.inspiracio.geometry.{Piclet, Point2}

import scala.swing._
import scala.swing.event.{MouseDragged, MousePressed, MouseReleased}

// Referenced classes of package bunkenba.calculator:
//            DoubleBuffer, World, Drawing
/** Representation of a world of complex numbers:
  * as Cartesian plane or as Riemann sphere. */
abstract class WorldRepresentation protected(w: World) extends Component {
  background = Color.white
  //doubleBuffered = true

  protected val short = new ComplexFormat
  short.maximumFractionDigits = 3

  //Methods ------------------------------------------------------

  private[calculator] def draw(g: Graphics, z: Complex)
  private[calculator] def draw(g: Graphics, zs: List[Complex] )
  private[calculator] def draw(g: Graphics, p: Piclet): Unit = draw(g, p.samples)

  override def preferredSize: Dimension = minimumSize

  /** Initial size wide enough so that the interaction menu is visible. */
  override def minimumSize: Dimension = new Dimension(550, 300)

  private[calculator] def point2Complex(p: Point): Option[Complex]

  private[calculator] def reset()

  /** shift calls repaint */
  private[calculator] def startShift(start: Point) = {}
  private[calculator] def shift(start: Point, from: Point, to: Point)
  private[calculator] def endShift(start: Point, end: Point) = {}

  //There is zoom for plane but not for sphere. Zoom causes repaint.
  private[calculator] def zoomIn(): Unit = {}
  private[calculator] def zoomOut(): Unit = {}
  private[calculator] def zoom: Double = 0
  private[calculator] def zoom_=(z: Double) : Unit = {}

  // helpers -----------------------------------

  protected[calculator] def pairs[A](cs : List[A]): List[(A,A)] = cs match {
    case Nil => Nil
    case a::Nil => Nil
    case a::b::cs => (a,b) :: pairs(b::cs)
  }

  protected def width = size.width
  protected def height = size.height

  listenTo(mouse.clicks)
  listenTo(mouse.moves)
  reactions += {
    case MousePressed(source, point, modifiers, clicks, triggersPopup) => w.mousePressed(point)
    case MouseDragged(source, point, modifiers) => w.mouseDragged(point)
    case MouseReleased(source, point, modifiers, clicks, triggersPopup) => w.mouseReleased(point)
  }

}